src/colmap/estimators/cost_functions.h

```diff
@@ -301,6 +302,111 @@ class SampsonErrorCostFunction {
+// splitting SE(3) into SO(3) x R^3. The 6x6 covariance matrix is defined in the
+// reference frame of the camera. Its first and last three components correspond
+// to the rotation and translation errors, respectively.
+struct AbsolutePoseErrorCostFunction {
+ public:
+  explicit AbsolutePoseErrorCostFunction(const Rigid3d& cam_from_world,
+                                         const EigenMatrix6d& covariance)
+      : cam_from_world_(cam_from_world),
+        sqrt_information_(covariance.inverse().llt().matrixL()) {}
+
+  static ceres::CostFunction* Create(const Rigid3d& cam_from_world,
+                                     const EigenMatrix6d& covariance) {
+    return (
+        new ceres::AutoDiffCostFunction<AbsolutePoseErrorCostFunction, 6, 4, 3>(
+            new AbsolutePoseErrorCostFunction(cam_from_world, covariance)));
+  }
+
+  template <typename T>
+  bool operator()(const T* const cam_from_world_q,
+                  const T* const cam_from_world_t,
+                  T* residuals_ptr) const {
+    const Eigen::Quaternion<T> mes_from_est_q =
+        cam_from_world_.rotation.cast<T>() *
+        EigenQuaternionMap<T>(cam_from_world_q).inverse();
+    EigenQuaternionToAngleAxis(mes_from_est_q.coeffs().data(), residuals_ptr);
+
+    Eigen::Map<Eigen::Matrix<T, 3, 1>> mes_from_est_t(residuals_ptr + 3);
+    mes_from_est_t = cam_from_world_.translation.cast<T>() -
+                     mes_from_est_q * EigenVector3Map<T>(cam_from_world_t);
+
+    Eigen::Map<Eigen::Matrix<T, 6, 1>> residuals(residuals_ptr);
+    residuals.applyOnTheLeft(sqrt_information_.template cast<T>());
+    return true;
+  }
+
+ private:
+  const Rigid3d& cam_from_world_;
+  const EigenMatrix6d sqrt_information_;
+};
+
+// 6-DoF error between two absolute poses based on a measurement that is their
+// relative pose, with identical scale for the translation. The covariance is
+// defined in the reference frame of the camera j.
+// Its first and last three components correspond to the rotation and
+// translation errors, respectively.
+struct MetricRelativePoseErrorCostFunction {
+ public:
+  explicit MetricRelativePoseErrorCostFunction(const Rigid3d& j_from_i,
+                                               const EigenMatrix6d& covariance)
+      : j_from_i_(j_from_i),
+        sqrt_information_(covariance.inverse().llt().matrixL()) {}
+
+  static ceres::CostFunction* Create(const Rigid3d& j_from_i,
+                                     const EigenMatrix6d& covariance) {
+    return (new ceres::AutoDiffCostFunction<MetricRelativePoseErrorCostFunction,
+                                            6,
+                                            4,
+                                            3,
+                                            4,
+                                            3>(
+        new MetricRelativePoseErrorCostFunction(j_from_i, covariance)));
+  }
+
+  template <typename T>
+  bool operator()(const T* const i_from_world_q,
```

Comment:
The following is probably more efficient, as inverting a quaternion is much cheaper as compared to applying it to a point:
```
template <typename T>
  bool operator()(const T* const i_from_world_q,
                  const T* const i_from_world_t,
                  const T* const j_from_world_q,
                  const T* const j_from_world_t,
                  T* residuals_ptr) const {
    const Eigen::Quaternion<T> j_from_i_q =
        EigenQuaternionMap<T>(j_from_world_q) *
        EigenQuaternionMap<T>(i_from_world_q).inverse();
    Eigen::Matrix<T, 3, 1> j_from_i_t =
        EigenVector3Map<T>(j_from_world_t) -
        j_from_i_q * EigenVector3Map<T>(i_from_world_t);

    const Eigen::Quaternion<T> mes_from_est_q =
        j_from_i_.rotation.cast<T>().inverse() * j_from_i_q;
    EigenQuaternionToAngleAxis(mes_from_est_q.coeffs().data(), residuals_ptr);

    Eigen::Map<Eigen::Matrix<T, 3, 1>> mes_from_est_t(residuals_ptr + 3);
    mes_from_est_t = j_from_i_.translation.cast<T>() - j_from_i_t;

    Eigen::Map<Eigen::Matrix<T, 6, 1>> residuals(residuals_ptr);
    residuals.applyOnTheLeft(sqrt_information_.template cast<T>());
    return true;
  }
```

Reply:
Same comment applies to the absolute pose error cost function above. If you compute j_from_i_q instead of i_from_j_q, you can save some compute.

Reply:
I think that this isn't mathematically correct. If we write ${}_j\hat{T}_i$ the measured transform and ${}_iT_w$ and ${}_jT_w$ the parameters, the error in the reference frame $j$ of the measurement is
$$\Delta T = {}_j\hat{T}_i \cdot {}_iT_w \cdot {}_jT_w^{-1}\enspace.$$
For the translation, this amounts to
$$\Delta t = {}_j\hat{t}_i + {}_j\hat{R}_i \cdot {}_it_j\enspace.$$
This is basically the center of $j$ (parameter) in the reference frame of the measure $j$. What you wrote amounts to:
$$\Delta t = {}_j\hat{t}_i + {}_jR_i \cdot {}_it_j\enspace.$$
This makes the simplification ${}_j\hat{R}_i \approx {}_jR_i\ $, which does not necessarily hold.

Reply:
Sorry, on my phone right now, and cannot right down the math. It looks like that in your notation my code does the following:
```
$$\Delta T = {}_jT_w \cdot {}_iT_w^{-1}  \cdot {}_j\hat{T}_i$$
```

Reply:
Reconsidering what I stated above: since Ceres use the left-hand convention for the quaternion retractation, the residual should indeed be as you describe. I'll update the code accordingly.

Reply:
Updated. This doesn't match your code (because of the approximation explained above) but is still efficient by expecting the constraint as ${}_i\hat{T}_j$ instead of ${}_j\hat{T}_i$. Also saved an inversion in `AbsolutePoseErrorCostFunction` by pre-computing the inverse.

Reply:
LGTM, thx